db/keyval/filesystem/client.go

```diff
@@ -0,0 +1,355 @@
+	}
+
+	// Do the initial read and store everything to internal database
+	for _, path := range paths {
+		// Register to watcher
+		watcher.Add(path)
+
+		isDir, err := c.r.IsDirectory(path)
+		if err != nil {
+			return nil, err
+		}
+		// Read file, or all files if path is a directory
+		var filesInPath []reader.File
+		if isDir {
+			filesInPath, err = c.r.ProcessFilesInDir(path)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			fileInPath, err := c.r.ProcessFile(path)
+			if err != nil {
+				return nil, err
+			}
+			filesInPath = []reader.File{fileInPath}
+		}
+		// Find all the configuration for given agent and key and store it in the database
+		for _, file := range filesInPath {
+			for _, fileData := range file.Data {
+				if ok := c.checkAgentPrefix(fileData.Key); ok {
+					c.db.Add(file.Path, fileData.Key, fileData.Value)
+				}
+			}
+		}
+	}
+
+	return c, nil
+}
+
+// GetPaths returns client file paths
+func (c *Client) GetPaths() []string {
+	return c.paths
+}
+
+// GetPrefix returns current agent prefix
+func (c *Client) GetPrefix() string {
+	return c.agentPrefix
+}
+
+// GetDB returns filesystem database
+func (c *Client) GetDB() FilesSystemDB {
+	return c.db
+}
+
+// GetWatcher returns filesystem watcher
+func (c *Client) GetWatcher() *fsnotify.Watcher {
+	return c.watcher
+}
+
+// Represents data with both identifiers, file path and key.
+type keyedData struct {
+	path string
+	watchResp
+}
+
+// BrokerWatcher implements CoreBrokerWatcher and provides broker/watcher constructors with client
+type BrokerWatcher struct {
+	*Client
+	prefix string
+}
+
+// NewBroker provides BytesBroker object with client and given prefix
+func (c *Client) NewBroker(prefix string) keyval.BytesBroker {
+	return &BrokerWatcher{
+		Client: c,
+		prefix: prefix,
+	}
+}
+
+// NewWatcher provides BytesWatcher object with client and given prefix
+func (c *Client) NewWatcher(prefix string) keyval.BytesWatcher {
+	return &BrokerWatcher{
+		Client: c,
+		prefix: prefix,
+	}
+}
+
+// Put is not supported, filesystem plugin does not allow to do changes to the configuration
+func (c *Client) Put(key string, data []byte, opts ...datasync.PutOption) error {
+	c.log.Warnf("adding configuration to filesystem is currently not allowed")
+	return nil
+}
+
+// NewTxn is not supported, filesystem plugin does not allow to do changes to the configuration
+func (c *Client) NewTxn() keyval.BytesTxn {
+	c.log.Warnf("creating transaction chains in filesystem is currently not allowed")
+	return nil
+}
+
+// GetValue returns a value for given key
+func (c *Client) GetValue(key string) (data []byte, found bool, revision int64, err error) {
+	data, found = c.db.GetDataForKey(c.agentPrefix + key)
+	return
+}
+
+// ListValues returns a list of values for given prefix
+func (c *Client) ListValues(prefix string) (keyval.BytesKeyValIterator, error) {
+	keyValues := c.db.GetValuesForPrefix(c.agentPrefix + prefix)
+	data := make([]*reader.FileEntry, 0)
```

Comment:
You know the final capacity here, you can initialize the slice with `make([]*reader.FileEntry, 0, len(keyValues))` to prevent needs for reallocating the underlying array.